fs/fs.go

```diff
@@ -117,14 +117,27 @@ func NewFsInfo(context Context) (FsInfo, error) {
 
 func (self *RealFsInfo) addLabels(context Context) {
 	dockerPaths := getDockerImagePaths(context)
+
+	// Determine if Docker is using devicemapper for its storage driver. If so, don't invoke
+	// self.updateDockerImagesPath below, as that will incorrectly override the device associated with
+	// LabelDockerImages (set previously up above in NewFsInfo).
+	dockerUsesDeviceMapper := false
+	for dev, p := range self.partitions {
+		if p.fsType == "devicemapper" && self.labels[LabelDockerImages] == dev {
```

Comment:
But refactoring these checks in combination with the check on line 373 that should work.

Problem I can see with the way you've currently done it is if anyone mounts a non-Docker devicemapper partition with Docker running on non-devicemapper storage things won't be labelled properly. A niche case of course but you know someone out there's doing it :)